sw/device/tests/otbn/otbn_ecdsa_p256_test.c

```diff
@@ -0,0 +1,262 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_otbn.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/otbn.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+/**
+ * ECDSA sign and verify test with the NIST P-256 curve using OTBN.
+ *
+ * IMPORTANT: This test is not a secure, complete, or reusable implementation of
+ * a cryptographic algorithm; it is not even close to being production-ready.
+ * It is only meant as an end-to-end test for OTBN during the bringup phase.
+ *
+ * The test contains constants and expected output, which can be independently
+ * and conveniently verified using a Python script.
+ *
+ * <code>
+ * # Optional: generate a new key
+ * $ openssl ecparam -name prime256v1 -genkey -noout -out \
+ *     otbn_ecdsa_p256_test_private_key.pem
+ *
+ * # Create all constants/variables
+ * $ ./otbn_test_params.py ecc otbn_ecdsa_p256_test_private_key.pem
+ * </code>
+ */
+
+OTBN_DECLARE_APP_SYMBOLS(p256_ecdsa);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, p256_ecdsa_sign);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, p256_ecdsa_verify);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, k);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, rnd);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, msg);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, r);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, s);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, x);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, y);
+OTBN_DECLARE_PTR_SYMBOL(p256_ecdsa, d);
+
+static const otbn_app_t kOtbnAppP256Ecdsa = OTBN_APP_T_INIT(p256_ecdsa);
+static const otbn_ptr_t kOtbnAppP256EcdsaFuncSign =
+    OTBN_PTR_T_INIT(p256_ecdsa, p256_ecdsa_sign);
+static const otbn_ptr_t kOtbnAppP256EcdsaFuncVerify =
+    OTBN_PTR_T_INIT(p256_ecdsa, p256_ecdsa_verify);
+static const otbn_ptr_t kOtbnVarK = OTBN_PTR_T_INIT(p256_ecdsa, k);
+static const otbn_ptr_t kOtbnVarRnd = OTBN_PTR_T_INIT(p256_ecdsa, rnd);
+static const otbn_ptr_t kOtbnVarMsg = OTBN_PTR_T_INIT(p256_ecdsa, msg);
+static const otbn_ptr_t kOtbnVarR = OTBN_PTR_T_INIT(p256_ecdsa, r);
+static const otbn_ptr_t kOtbnVarS = OTBN_PTR_T_INIT(p256_ecdsa, s);
+static const otbn_ptr_t kOtbnVarX = OTBN_PTR_T_INIT(p256_ecdsa, x);
+static const otbn_ptr_t kOtbnVarY = OTBN_PTR_T_INIT(p256_ecdsa, y);
+static const otbn_ptr_t kOtbnVarD = OTBN_PTR_T_INIT(p256_ecdsa, d);
+
+const test_config_t kTestConfig;
+
+/**
+ * CHECK()s that the actual data matches the expected data.
+ *
+ * @param actual The actual data.
+ * @param expected The expected data.
+ * @param size_bytes The size of the actual/expected data.
+ */
+static void check_data(const uint8_t *actual, const uint8_t *expected,
```

Comment:
Can you add a message parameter? Just to make it quicker to see what is broken.

```c
static void check_data(const char *msg, const uint8_t *actual, const uint8_t *expected, size_t size_bytes) {
  ...
    CHECK(actual[i] == expected[i],
          "%s: mismatch at byte %d: 0x%x (actual) != 0x%x (expected)", msg, i,
          actual[i], expected[i]);
```
Call site:
```c
check_data("signature_r", signature_r, kExpectedSignatureR, 32);
```